File: admin/src/lib/customer-order-status.ts
import { DeliveryType, OrderStatus } from "@prisma/client";

/**
 * مراحل مبسّطة ظاهرة للعميل فقط.
 * الحالات الداخلية للمطبخ/الأدمن تبقى كما هي عبر getOrderStatusLabel.
 */
export type CustomerOrderStage =
  | "AWAITING_KITCHEN"
  | "PAY_DEPOSIT"
  | "DEPOSIT_REVIEW"
  | "PREPARING"
  | "ON_THE_WAY"
  | "CONFIRM_RECEIPT"
  | "COMPLETED"
  | "REJECTED"
  | "CANCELLED";

export function getCustomerOrderStage(status: OrderStatus): CustomerOrderStage {
  switch (status) {
    case OrderStatus.PENDING_KITCHEN_APPROVAL:
      return "AWAITING_KITCHEN";
    case OrderStatus.ACCEPTED_AWAITING_DEPOSIT:
      return "PAY_DEPOSIT";
    case OrderStatus.DEPOSIT_PROOF_SUBMITTED:
      return "DEPOSIT_REVIEW";
    case OrderStatus.DEPOSIT_CONFIRMED:
    case OrderStatus.PREPARING:
      return "PREPARING";
    case OrderStatus.READY_FOR_PICKUP:
    case OrderStatus.AWAITING_CUSTOMER_ARRIVAL:
    case OrderStatus.OUT_FOR_DELIVERY:
      return "ON_THE_WAY";
    case OrderStatus.DELIVERED_BY_KITCHEN_OR_DRIVER:
    case OrderStatus.COMPLETED_AWAITING_CUSTOMER_CONFIRM:
      return "CONFIRM_RECEIPT";
    case OrderStatus.COMPLETED:
      return "COMPLETED";
    case OrderStatus.REJECTED_BY_KITCHEN:
      return "REJECTED";
    case OrderStatus.CANCELLED_BEFORE_DEPOSIT:
    case OrderStatus.CANCELLED_AFTER_DEPOSIT:
      return "CANCELLED";
  }
}

export function getCustomerOrderStatusLabel(
  status: OrderStatus,
  deliveryType: DeliveryType = DeliveryType.DELIVERY,
): string {
  switch (getCustomerOrderStage(status)) {
    case "AWAITING_KITCHEN":
      return "بانتظار موافقة المطبخ";
    case "PAY_DEPOSIT":
      return "ادفع العربون";
    case "DEPOSIT_REVIEW":
      return "بانتظار تأكيد العربون";
    case "PREPARING":
      return "جاري التحضير";
    case "ON_THE_WAY":
      return deliveryType === DeliveryType.PICKUP
        ? "جاهز للاستلام"
        : "في الطريق إليك";
    case "CONFIRM_RECEIPT":
      return "أكّد استلام الطلب";
    case "COMPLETED":
      return "مكتمل";
    case "REJECTED":
      return "مرفوض من المطبخ";
    case "CANCELLED":
      return "تم الإلغاء";
  }
}

export function getCustomerOrderStatusHint(
  status: OrderStatus,
  deliveryType: DeliveryType = DeliveryType.DELIVERY,
): string {
  switch (getCustomerOrderStage(status)) {
    case "AWAITING_KITCHEN":
      return "المطبخ هيراجع الطلب ويحدّد رسوم التوصيل إن وُجدت.";
    case "PAY_DEPOSIT":
      return "حوّل العربون وأرسل صورة الإثبات عشان يبدأ التحضير.";
    case "DEPOSIT_REVIEW":
      return "المطبخ بيراجع إثبات العربون.";
    case "PREPARING":
      return "المطبخ بيجهّز طلبك.";
    case "ON_THE_WAY":
      return deliveryType === DeliveryType.PICKUP
        ? "الطلب جاهز — توجّه لاستلامه من المطبخ."
        : "الطلب في الطريق لعنوانك.";
    case "CONFIRM_RECEIPT":
      return "ادفع الباقي عند الاستلام، ثم أكّد إنك استلمت الطلب.";
    case "COMPLETED":
      return "تم إنهاء الطلب. تقدر تقيّم المطبخ.";
    case "REJECTED":
      return "المطبخ رفض الطلب. تقدر تطلب من مطبخ آخر.";
    case "CANCELLED":
      return "تم إلغاء هذا الطلب.";
  }
}

/** العميل يلغي فقط قبل تأكيد العربون. */
export function canCustomerCancelOrder(status: OrderStatus): boolean {
  return (
    status === OrderStatus.PENDING_KITCHEN_APPROVAL ||
    status === OrderStatus.ACCEPTED_AWAITING_DEPOSIT ||
    status === OrderStatus.DEPOSIT_PROOF_SUBMITTED
  );
}

File: takka_mobile/lib/core/orders/customer_order_status.dart
/// مراحل مبسّطة ظاهرة للعميل فقط.
/// الحالات الداخلية للمطبخ تبقى كما هي عبر [orderStatusLabelAr].

enum CustomerOrderStage {
  awaitingKitchen,
  payDeposit,
  depositReview,
  preparing,
  onTheWay,
  confirmReceipt,
  completed,
  rejected,
  cancelled,
}

CustomerOrderStage customerOrderStage(String status) {
  switch (status) {
    case 'PENDING_KITCHEN_APPROVAL':
      return CustomerOrderStage.awaitingKitchen;
    case 'ACCEPTED_AWAITING_DEPOSIT':
      return CustomerOrderStage.payDeposit;
    case 'DEPOSIT_PROOF_SUBMITTED':
      return CustomerOrderStage.depositReview;
    case 'DEPOSIT_CONFIRMED':
    case 'PREPARING':
      return CustomerOrderStage.preparing;
    case 'READY_FOR_PICKUP':
    case 'AWAITING_CUSTOMER_ARRIVAL':
    case 'OUT_FOR_DELIVERY':
      return CustomerOrderStage.onTheWay;
    case 'DELIVERED_BY_KITCHEN_OR_DRIVER':
    case 'COMPLETED_AWAITING_CUSTOMER_CONFIRM':
      return CustomerOrderStage.confirmReceipt;
    case 'COMPLETED':
      return CustomerOrderStage.completed;
    case 'REJECTED_BY_KITCHEN':
      return CustomerOrderStage.rejected;
    case 'CANCELLED_BEFORE_DEPOSIT':
    case 'CANCELLED_AFTER_DEPOSIT':
      return CustomerOrderStage.cancelled;
    default:
      return CustomerOrderStage.awaitingKitchen;
  }
}

String customerOrderStatusLabel(
  String status, {
  String deliveryType = 'DELIVERY',
}) {
  switch (customerOrderStage(status)) {
    case CustomerOrderStage.awaitingKitchen:
      return 'بانتظار موافقة المطبخ';
    case CustomerOrderStage.payDeposit:
      return 'ادفع العربون';
    case CustomerOrderStage.depositReview:
      return 'بانتظار تأكيد العربون';
    case CustomerOrderStage.preparing:
      return 'جاري التحضير';
    case CustomerOrderStage.onTheWay:
      return deliveryType == 'PICKUP' ? 'جاهز للاستلام' : 'في الطريق إليك';
    case CustomerOrderStage.confirmReceipt:
      return 'أكّد استلام الطلب';
    case CustomerOrderStage.completed:
      return 'مكتمل';
    case CustomerOrderStage.rejected:
      return 'مرفوض من المطبخ';
    case CustomerOrderStage.cancelled:
      return 'تم الإلغاء';
  }
}

String customerOrderStatusHint(
  String status, {
  String deliveryType = 'DELIVERY',
}) {
  switch (customerOrderStage(status)) {
    case CustomerOrderStage.awaitingKitchen:
      return 'المطبخ هيراجع الطلب ويحدّد رسوم التوصيل إن وُجدت.';
    case CustomerOrderStage.payDeposit:
      return 'حوّل العربون وأرسل صورة الإثبات عشان يبدأ التحضير.';
    case CustomerOrderStage.depositReview:
      return 'المطبخ بيراجع إثبات العربون.';
    case CustomerOrderStage.preparing:
      return 'المطبخ بيجهّز طلبك.';
    case CustomerOrderStage.onTheWay:
      return deliveryType == 'PICKUP'
          ? 'الطلب جاهز — توجّه لاستلامه من المطبخ.'
          : 'الطلب في الطريق لعنوانك.';
    case CustomerOrderStage.confirmReceipt:
      return 'ادفع الباقي عند الاستلام، ثم أكّد إنك استلمت الطلب.';
    case CustomerOrderStage.completed:
      return 'تم إنهاء الطلب. تقدر تقيّم المطبخ.';
    case CustomerOrderStage.rejected:
      return 'المطبخ رفض الطلب. تقدر تطلب من مطبخ آخر.';
    case CustomerOrderStage.cancelled:
      return 'تم إلغاء هذا الطلب.';
  }
}

/// العميل يلغي فقط قبل تأكيد العربون.
bool canCustomerCancelOrder(String status) {
  return status == 'PENDING_KITCHEN_APPROVAL' ||
      status == 'ACCEPTED_AWAITING_DEPOSIT' ||
      status == 'DEPOSIT_PROOF_SUBMITTED';
}
